# Lesson 2

## Working with requests library

Now we will start doing a request to the following websites: Google, NBA and Rotten Tomatoes. 

Let's check their status_code by using the requests method GET, which simply indicates that you're trying to retrieve information from the data source:

In [1]:
import requests

google = requests.get("https://developers.google.com")
print("Google:", google.status_code)

NBA = response = requests.get("https://api.sportsdata.io/api/nba/fantasy/json/CurrentSeason")
print("NBA:", NBA.status_code)

rotten_tomato = requests.get("http://api.rottentomatoes.com/api/public/v1.0/lists/movies/box_office.json")
print("Rotten Tomatoes:", rotten_tomato.status_code)

Google: 200
NBA: 401
Rotten Tomatoes: 403


## Status codes

So we can see we receive some status codes, the most frequent are these:

* **200**: Everything went okay and the result has been returned (if any).

* **301**: The server is redirecting you to a different endpoint. This can happen when a company switches domain names, or an endpoint name is changed.

* **400**: The server thinks you made a bad request. This happens when you don’t send along the right data, among other things.

* **401**: You are not properly authenticated.

* **403**: The resource you’re trying to access is forbidden: you don’t have the right permissions to get it.

* **404**: The resource you tried to access doesn't exist.

* **503**: The server can't handle the request.

## After connecting

As we are connected now to the google developer endpoint let's try to get some data from it.

In [2]:
print(google.text)

<!doctype html>
<html 
      lang="en"
      dir="ltr">
  <head>
    
    
    
    
    <meta name="google-signin-client-id" content="721724668570-nbkv1cfusk7kk4eni4pjvepaus73b13t.apps.googleusercontent.com">
    <meta name="google-signin-scope"
          content="profile email">
    <meta property="og:site_name" content="Google Developers">
    <meta property="og:type" content="website">
    

    
      <meta name="theme-color" content="#ffffff">
    

    <meta charset="utf-8">
    <meta content="IE=Edge" http-equiv="X-UA-Compatible">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="manifest" href="/_pwa/developers/manifest.json"
          crossorigin="use-credentials">
    <link rel="preconnect" href="//www.gstatic.com" crossorigin>
    <link rel="preconnect" href="//fonts.gstatic.com" crossorigin>
    <link rel="preconnect" href="//fonts.googleapis.com" crossorigin>
    <link rel="preconnect" href="//apis.google.com" crossorigin>
    <link r

This is not the proper format for us right now so let's try to tranform it to the one we need, the **JSON (JavaScript Object Notation) format**, which is the language of RESTful APIs. It is the way to encode data to make data easily readable by machines.

In [3]:
#google.json()

The problem here is that we are not connected to a proper API that has a json as the endpoint, so we will need to get another valid URL. Remember that the code 200 means that the connection has been established, but not that it is the connection you need. Let's try with the skyscrapper API .

We will connect to this API through the web [RapidAPI ](https://rapidapi.com/skyscanner/api/skyscanner-flight-search)so you will need to create an account to get a personal Key (user ID) to have access to it, just sign up and get your Key. The "problem" with this API is that we will need to pass some parameters, including the headers and the city following this code:

In [6]:
#Look into this
#from google.colab import drive
#drive.mount('/content/drive')

In [10]:
#import sys
#sys.path.insert(1, '/content/drive/MyDrive/Credentials/')
#from config import *

In [11]:
#keys
x_rapidapi_key="eb056a72f9msh2564512806e6aebp12a874jsnd03d93885916"

In [12]:
import requests

url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/autosuggest/v1.0/UK/GBP/en-GB/"

params = {"query":"Tokyo"}

#headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
#                      'x-rapidapi-key': "<introduce your RapidAPI key here>"}

headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
                      'x-rapidapi-key': x_rapidapi_key}

# Notice the additional keywords here (headers, params)
response = requests.get(url, headers = headers, params = params)
response.json()

{'Places': [{'PlaceId': 'TYOA-sky',
   'PlaceName': 'Tokyo',
   'CountryId': 'JP-sky',
   'RegionId': '',
   'CityId': 'TYOA-sky',
   'CountryName': 'Japan'},
  {'PlaceId': 'NRT-sky',
   'PlaceName': 'Tokyo Narita',
   'CountryId': 'JP-sky',
   'RegionId': '',
   'CityId': 'TYOA-sky',
   'CountryName': 'Japan'},
  {'PlaceId': 'HND-sky',
   'PlaceName': 'Tokyo Haneda',
   'CountryId': 'JP-sky',
   'RegionId': '',
   'CityId': 'TYOA-sky',
   'CountryName': 'Japan'},
  {'PlaceId': 'TJH-sky',
   'PlaceName': 'Toyooka',
   'CountryId': 'JP-sky',
   'RegionId': '',
   'CityId': 'JTJH-sky',
   'CountryName': 'Japan'},
  {'PlaceId': 'OOK-sky',
   'PlaceName': 'Toksook Bay',
   'CountryId': 'US-sky',
   'RegionId': 'AK',
   'CityId': 'OOKA-sky',
   'CountryName': 'United States'},
  {'PlaceId': 'TKZ-sky',
   'PlaceName': 'Tokoroa',
   'CountryId': 'NZ-sky',
   'RegionId': '',
   'CityId': 'TKZN-sky',
   'CountryName': 'New Zealand'}]}

You should get:

{"Places":[{"PlaceId":"TYOA-sky","PlaceName":"Tokyo","CountryId":"JP-sky","RegionId":"","CityId":"TYOA-sky","CountryName":"Japan"},{"PlaceId":"NRT-sky","PlaceName":"Tokyo Narita","CountryId":"JP-sky","RegionId":"","CityId":"TYOA-sky","CountryName":"Japan"},{"PlaceId":"HND-sky","PlaceName":"Tokyo Haneda","CountryId":"JP-sky","RegionId":"","CityId":"TYOA-sky","CountryName":"Japan"},{"PlaceId":"TJH-sky","PlaceName":"Toyooka","CountryId":"JP-sky","RegionId":"","CityId":"JTJH-sky","CountryName":"Japan"},{"PlaceId":"OOK-sky","PlaceName":"Toksook Bay","CountryId":"US-sky","RegionId":"AK","CityId":"OOKA-sky","CountryName":"United States"},{"PlaceId":"TKZ-sky","PlaceName":"Tokoroa","CountryId":"NZ-sky","RegionId":"","CityId":"TKZN-sky","CountryName":"New Zealand"}]}

To know the parameters we need to be able to connect properly to an API we should look first at its documentation. For the next query we will look at the prices for the flights from San Francisco to New York City in 2020/12/12

In [13]:
url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/browsequotes/v1.0/US/USD/en-US/SFO-sky/NYCA-sky/2020-12-12"

params = {"inboundpartialdate":"2020-12-12"}

#headers = {
#    'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
#    'x-rapidapi-key': "<introduce your RapidAPI key here>"}

headers = {
    'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
    'x-rapidapi-key': x_rapidapi_key}    

response = requests.get(url, headers=headers, params=params)

response.json()

{'Quotes': [{'QuoteId': 1,
   'MinPrice': 62,
   'Direct': False,
   'OutboundLeg': {'CarrierIds': [1065],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-11-23T21:49:00'},
  {'QuoteId': 2,
   'MinPrice': 85,
   'Direct': True,
   'OutboundLeg': {'CarrierIds': [851],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-11-23T21:49:00'}],
 'Carriers': [{'CarrierId': 1065, 'Name': 'Frontier Airlines'},
  {'CarrierId': 851, 'Name': 'Alaska Airlines'}],
 'Places': [{'Name': 'New York Newark',
   'Type': 'Station',
   'PlaceId': 50290,
   'IataCode': 'EWR',
   'SkyscannerCode': 'EWR',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'Name': 'San Francisco International',
   'Type': 'Station',
   'PlaceId': 81727,
   'IataCode': 'SFO',
   'SkyscannerCode': 'SFO',
   'CityName': 'San Francisco',
   'CityId': 'SFOA'

The output will be:

{'Quotes': [{'QuoteId': 1,
   'MinPrice': 92.0,
   'Direct': False,
   'OutboundLeg': {'CarrierIds': [1065],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-09-21T10:49:00'},
  {'QuoteId': 2,
   'MinPrice': 133.0,
   'Direct': True,
   'OutboundLeg': {'CarrierIds': [851],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-09-21T10:49:00'}],
 'Places': [{'PlaceId': 50290,
   'IataCode': 'EWR',
   'Name': 'New York Newark',
   'Type': 'Station',
   'SkyscannerCode': 'EWR',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 60987,
   'IataCode': 'JFK',
   'Name': 'New York John F. Kennedy',
   'Type': 'Station',
   'SkyscannerCode': 'JFK',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 65633,
   'IataCode': 'LGA',
   'Name': 'New York LaGuardia',
   'Type': 'Station',
   'SkyscannerCode': 'LGA',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 81727,
   'IataCode': 'SFO',
   'Name': 'San Francisco International',
   'Type': 'Station',
   'SkyscannerCode': 'SFO',
   'CityName': 'San Francisco',
   'CityId': 'SFOA',
   'CountryName': 'United States'}],
 'Carriers': [{'CarrierId': 819, 'Name': 'Aegean Airlines'},
  {'CarrierId': 851, 'Name': 'Alaska Airlines'},
  {'CarrierId': 870, 'Name': 'jetBlue'},
  {'CarrierId': 1065, 'Name': 'Frontier Airlines'},
  {'CarrierId': 1721, 'Name': 'Sun Country Airlines'},
  {'CarrierId': 1793, 'Name': 'United'},
  {'CarrierId': 1902, 'Name': 'Southwest Airlines'}],
 'Currencies': [{'Code': 'USD',
   'Symbol': '$',
   'ThousandsSeparator': ',',
   'DecimalSeparator': '.',
   'SymbolOnLeft': True,
   'SpaceBetweenAmountAndSymbol': False,
   'RoundingCoefficient': 0,
   'DecimalDigits': 2}]}

# 8.04 Activity 2

Connect to the **SkyScanner** API and extract information from all the flights flying from Munich to Berlin. To do that you will need to get their _city ID_ which you can find in the first query we used in the lesson. In this activity you will just need to create a function called `city_code` that takes a `city_name`(_string-like_, _"Mumbai"_) as a parameter and returns the city flight code.



In [19]:
import requests

url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/autosuggest/v1.0/UK/GBP/en-GB/"

params = {"query":"Munich"}

#headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
#                      'x-rapidapi-key': "<introduce your RapidAPI key here>"}

headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
                      'x-rapidapi-key': "eb056a72f9msh2564512806e6aebp12a874jsnd03d93885916"}

# Notice the additional keywords here (headers, params)
response = requests.get(url, headers = headers, params = params)
response.json()

{'Places': [{'PlaceId': 'MUC-sky',
   'PlaceName': 'Munich',
   'CountryId': 'DE-sky',
   'RegionId': '',
   'CityId': 'MUNI-sky',
   'CountryName': 'Germany'},
  {'PlaceId': 'LGB-sky',
   'PlaceName': 'Long Beach Municipal',
   'CountryId': 'US-sky',
   'RegionId': 'CA',
   'CityId': 'LGBA-sky',
   'CountryName': 'United States'},
  {'PlaceId': 'SBA-sky',
   'PlaceName': 'Santa Barbara Municipal',
   'CountryId': 'US-sky',
   'RegionId': 'CA',
   'CityId': 'SBAA-sky',
   'CountryName': 'United States'},
  {'PlaceId': 'CAT-sky',
   'PlaceName': 'Cascais Municipal',
   'CountryId': 'PT-sky',
   'RegionId': '',
   'CityId': 'CATC-sky',
   'CountryName': 'Portugal'},
  {'PlaceId': 'RST-sky',
   'PlaceName': 'Rochester Municipal',
   'CountryId': 'US-sky',
   'RegionId': 'MN',
   'CityId': 'RSTA-sky',
   'CountryName': 'United States'},
  {'PlaceId': 'MOB-sky',
   'PlaceName': 'Mobile Municipal Arpt',
   'CountryId': 'US-sky',
   'RegionId': 'AL',
   'CityId': 'MOBA-sky',
   'CountryName':

In [33]:
def city_code(x):
    import requests

    url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/autosuggest/v1.0/UK/GBP/en-GB/"

    params = {"query":x}

    #headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
    #                      'x-rapidapi-key': "<introduce your RapidAPI key here>"}

    headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
                      'x-rapidapi-key': "eb056a72f9msh2564512806e6aebp12a874jsnd03d93885916"}

    # Notice the additional keywords here (headers, params)
    response = requests.get(url, headers = headers, params = params)
    response.json()
    
    return response.json()["Places"][0]["PlaceId"]

In [34]:
city_code("Munich")

'MUC-sky'

In [35]:
city_code("Berlin")

'BER-sky'

# Lesson 3 - DataFrames

## Handling JSON files

In [25]:
# Import libraries
import pandas as pd
import json

In [26]:
my_response = {'Quotes': [{'QuoteId': 1,
   'MinPrice': 92.0,
   'Direct': False,
   'OutboundLeg': {'CarrierIds': [1065],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-09-21T10:49:00'},
  {'QuoteId': 2,
   'MinPrice': 133.0,
   'Direct': True,
   'OutboundLeg': {'CarrierIds': [851],
    'OriginId': 81727,
    'DestinationId': 50290,
    'DepartureDate': '2020-12-12T00:00:00'},
   'QuoteDateTime': '2020-09-21T10:49:00'}],
 'Places': [{'PlaceId': 50290,
   'IataCode': 'EWR',
   'Name': 'New York Newark',
   'Type': 'Station',
   'SkyscannerCode': 'EWR',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 60987,
   'IataCode': 'JFK',
   'Name': 'New York John F. Kennedy',
   'Type': 'Station',
   'SkyscannerCode': 'JFK',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 65633,
   'IataCode': 'LGA',
   'Name': 'New York LaGuardia',
   'Type': 'Station',
   'SkyscannerCode': 'LGA',
   'CityName': 'New York',
   'CityId': 'NYCA',
   'CountryName': 'United States'},
  {'PlaceId': 81727,
   'IataCode': 'SFO',
   'Name': 'San Francisco International',
   'Type': 'Station',
   'SkyscannerCode': 'SFO',
   'CityName': 'San Francisco',
   'CityId': 'SFOA',
   'CountryName': 'United States'}],
 'Carriers': [{'CarrierId': 819, 'Name': 'Aegean Airlines'},
  {'CarrierId': 851, 'Name': 'Alaska Airlines'},
  {'CarrierId': 870, 'Name': 'jetBlue'},
  {'CarrierId': 1065, 'Name': 'Frontier Airlines'},
  {'CarrierId': 1721, 'Name': 'Sun Country Airlines'},
  {'CarrierId': 1793, 'Name': 'United'},
  {'CarrierId': 1902, 'Name': 'Southwest Airlines'}],
 'Currencies': [{'Code': 'USD',
   'Symbol': '$',
   'ThousandsSeparator': ',',
   'DecimalSeparator': '.',
   'SymbolOnLeft': True,
   'SpaceBetweenAmountAndSymbol': False,
   'RoundingCoefficient': 0,
   'DecimalDigits': 2}]}

In [27]:
pd.json_normalize(my_response)
# If this doesn't work, try to import json_normalize directly from pandas.io.json
#pd.DataFrame({'First_col': [], 'Second_col': [],....})

Quotes  \
0  [{'QuoteId': 1, 'MinPrice': 92.0, 'Direct': Fa...   

                                              Places  \
0  [{'PlaceId': 50290, 'IataCode': 'EWR', 'Name':...   

                                            Carriers  \
0  [{'CarrierId': 819, 'Name': 'Aegean Airlines'}...   

                                          Currencies  
0  [{'Code': 'USD', 'Symbol': '$', 'ThousandsSepa...

As an output we have a DataFrame with 4 columns, Quotes, Places, Carriers and Currencies, but the values will also be in JSON format. From this we can infere that we are not working with a dataset, but a relational database, and the columns are the datasets we were looking for. Let's transform them into pd.DataFrames:

In [28]:
quotes = pd.DataFrame(pd.json_normalize(my_response)["Quotes"][0])
carriers = pd.DataFrame(pd.json_normalize(my_response)["Carriers"][0])
places = pd.DataFrame(pd.json_normalize(my_response)["Places"][0])
currencies = pd.DataFrame(pd.json_normalize(my_response)["Currencies"][0])

print("Quotes DataFrame extract: ")
display(quotes.head())
print()
print("Carriers DataFrame extract: ")
display(carriers.head())
print()
print("Places DataFrame extract: ")
display(places.head())
print()
print("Currencies DataFrame extract: ")
display(currencies.head())

Quotes DataFrame extract: 


QuoteId  MinPrice  Direct  \
0        1      92.0   False   
1        2     133.0    True   

                                         OutboundLeg        QuoteDateTime  
0  {'CarrierIds': [1065], 'OriginId': 81727, 'Des...  2020-09-21T10:49:00  
1  {'CarrierIds': [851], 'OriginId': 81727, 'Dest...  2020-09-21T10:49:00


Carriers DataFrame extract: 


CarrierId                  Name
0        819       Aegean Airlines
1        851       Alaska Airlines
2        870               jetBlue
3       1065     Frontier Airlines
4       1721  Sun Country Airlines


Places DataFrame extract: 


PlaceId IataCode                         Name     Type SkyscannerCode  \
0    50290      EWR              New York Newark  Station            EWR   
1    60987      JFK     New York John F. Kennedy  Station            JFK   
2    65633      LGA           New York LaGuardia  Station            LGA   
3    81727      SFO  San Francisco International  Station            SFO   

        CityName CityId    CountryName  
0       New York   NYCA  United States  
1       New York   NYCA  United States  
2       New York   NYCA  United States  
3  San Francisco   SFOA  United States


Currencies DataFrame extract: 


Code Symbol ThousandsSeparator DecimalSeparator  SymbolOnLeft  \
0  USD      $                  ,                .          True   

   SpaceBetweenAmountAndSymbol  RoundingCoefficient  DecimalDigits  
0                        False                    0              2

After reviewing and printing our data we can conclude that quotes are the flights we have for our selected origin and destination (just one in this case), places are the airports we could travel from and to, carriers are the companies usually have the kind of trip we have chosen, and currency is self-explanatory.

However we still have another json in one of the datasets, in Quotes, variable ["OutboundLeg"], that's due to the same principle we have been working for during this lesson, relational databases. Let's unfold that variable:

In [41]:
#flights = pd.DataFrame(pd.DataFrame(pd.json_normalize(my_response)["Quotes"][0])["OutboundLeg"][0])
#flights.head()
#flights2 = pd.DataFrame(pd.json_normalize(my_response)["Quotes"][0])["OutboundLeg"][0]
pd.DataFrame(pd.DataFrame(pd.json_normalize(my_response)["Quotes"][0])["OutboundLeg"][0])

CarrierIds  OriginId  DestinationId        DepartureDate
0        1065     81727          50290  2020-12-12T00:00:00

# 8.04 Activity 3

Create a function called `flight_prices`that takes `departure` (as the origin city name, "Mumbai"), `arrival`(as the destination city name, "Istanbul") and `date`. It should return an output similar to the one we had for San Francisco - New York. You may want to check if it works with these param values:

- Departure: "Berlin"
- Arrival: "Vienna"
- Date: "2020-10-03" (The format will help but remember that it has to be an upcoming date)



In [42]:
def flight_prices(departure, arrival, date):

    departure_code = city_code(departure)
    arrival_code = city_code(arrival)
    url = f"https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/browsequotes/v1.0/US/USD/en-US/{departure_code}/{arrival_code}/{date}"

    params = {"inboundpartialdate":{date}}

    #headers = {
    #'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
    #'x-rapidapi-key': "<introduce your RapidAPI key here>"}

    headers = {
    'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
    'x-rapidapi-key': x_rapidapi_key}

    response = requests.get(url, headers=headers, params=params)

    return response.json()

In [43]:
flight_prices("Berlin","Vienna","2020-12-03")

{'Quotes': [{'QuoteId': 1,
   'MinPrice': 161,
   'Direct': False,
   'OutboundLeg': {'CarrierIds': [1755],
    'OriginId': 42514,
    'DestinationId': 88773,
    'DepartureDate': '2020-12-03T00:00:00'},
   'QuoteDateTime': '2020-11-24T11:57:00'},
  {'QuoteId': 2,
   'MinPrice': 177,
   'Direct': True,
   'OutboundLeg': {'CarrierIds': [1523],
    'OriginId': 42514,
    'DestinationId': 88773,
    'DepartureDate': '2020-12-03T00:00:00'},
   'QuoteDateTime': '2020-11-24T11:57:00'}],
 'Carriers': [{'CarrierId': 1755, 'Name': 'Turkish Airlines'},
  {'CarrierId': 1523, 'Name': 'Austrian Airlines'}],
 'Places': [{'Name': 'Berlin Brandenburg',
   'Type': 'Station',
   'PlaceId': 42514,
   'IataCode': 'BER',
   'SkyscannerCode': 'BER',
   'CityName': 'Berlin',
   'CityId': 'BERL',
   'CountryName': 'Germany'},
  {'Name': 'Vienna',
   'Type': 'Station',
   'PlaceId': 88773,
   'IataCode': 'VIE',
   'SkyscannerCode': 'VIE',
   'CityName': 'Vienna',
   'CityId': 'VIEN',
   'CountryName': 'Austria

# Lesson 4

Now that we know what JSON is and we have connected to the skyscanner API let's work with it. In this lesson we will create a function for which if we introduce two dates, an origin and a destination we will get the data from the cheapest flight between these two dates.

We will start by creating a function that takes two dates and return a list of dates between the given ones. Luckily, there is a pandas method for that.

In [30]:
def get_dates(start, end):
    return pd.Series(pd.date_range(start, end,freq='d').format()) # Format makes the output ans str variable.

get_dates('1992-01-01','1992-01-20')

0     1992-01-01
1     1992-01-02
2     1992-01-03
3     1992-01-04
4     1992-01-05
5     1992-01-06
6     1992-01-07
7     1992-01-08
8     1992-01-09
9     1992-01-10
10    1992-01-11
11    1992-01-12
12    1992-01-13
13    1992-01-14
14    1992-01-15
15    1992-01-16
16    1992-01-17
17    1992-01-18
18    1992-01-19
19    1992-01-20
dtype: object

In [31]:
def city_code(city_name):
    url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/autosuggest/v1.0/US/USD/en-US/"

    params = {"query": city_name}

    headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
                      'x-rapidapi-key': x_rapidapi_key}

    response = requests.get(url, headers = headers, params = params)

    return response.json()["Places"][0]["PlaceId"]

In [32]:
city_code('London')

'LOND-sky'

Let's use the funtion to get the flights between two destinations on a given date.

In [37]:
def flight_prices(departure, arrival, date):

    departure_code = city_code(departure)
    arrival_code = city_code(arrival)
    url = f"https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/browsequotes/v1.0/US/USD/en-US/{departure_code}/{arrival_code}/{date}"

    params = {"inboundpartialdate":{date}}

    #headers = {
    #'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
    #'x-rapidapi-key': "<introduce your RapidAPI key here>"}

    headers = {
    'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
    'x-rapidapi-key': x_rapidapi_key}

    response = requests.get(url, headers=headers, params=params)

    return response.json()

In [38]:
flight_prices('Paris', 'London','2020-12-10')

{'Quotes': [{'QuoteId': 1,
   'MinPrice': 61,
   'Direct': True,
   'OutboundLeg': {'CarrierIds': [881],
    'OriginId': 44759,
    'DestinationId': 65698,
    'DepartureDate': '2020-12-10T00:00:00'},
   'QuoteDateTime': '2020-11-24T11:03:00'}],
 'Carriers': [{'CarrierId': 881, 'Name': 'British Airways'}],
 'Places': [{'Name': 'Paris Charles de Gaulle',
   'Type': 'Station',
   'PlaceId': 44759,
   'IataCode': 'CDG',
   'SkyscannerCode': 'CDG',
   'CityName': 'Paris',
   'CityId': 'PARI',
   'CountryName': 'France'},
  {'Name': 'London Heathrow',
   'Type': 'Station',
   'PlaceId': 65698,
   'IataCode': 'LHR',
   'SkyscannerCode': 'LHR',
   'CityName': 'London',
   'CityId': 'LOND',
   'CountryName': 'United Kingdom'},
  {'Name': 'Paris',
   'Type': 'City',
   'PlaceId': 3579769,
   'SkyscannerCode': 'PARI',
   'CityId': 'PARI',
   'CountryName': 'France'}],
 'Currencies': [{'Code': 'USD',
   'Symbol': '$',
   'ThousandsSeparator': ',',
   'DecimalSeparator': '.',
   'SymbolOnLeft': Tr

Now that we already have our date function let's create another function that takes the origin, destination and dates as arguments. It will create a dictionary that has every day between the two dates and containts a call to the API with all the information about the days:

In [39]:
from  time import sleep

def flight_days(origin, destination, start, end):
    dates = get_dates(start, end)
    my_dict = {}
    for date in list(dates):
        result = None
        while (result == None):
            try: 
                print(date)
                print("Gathering fligths")
                result = flight_prices(origin, destination, date)
                my_dict[date] = result
                sleep(5)
            except: 
                print("Error from the API")
                sleep(60)
    return my_dict

flight_days('Paris','Barcelona','2020-11-26','2020-12-10')

2020-11-26
Gathering fligths
2020-11-27
Gathering fligths
2020-11-28
Gathering fligths
2020-11-29
Gathering fligths
2020-11-30
Gathering fligths
2020-12-01
Gathering fligths
2020-12-02
Gathering fligths
2020-12-03
Gathering fligths
2020-12-04
Gathering fligths
2020-12-05
Gathering fligths
2020-12-06
Gathering fligths
2020-12-07
Gathering fligths
2020-12-08
Gathering fligths
2020-12-09
Gathering fligths
2020-12-10
Gathering fligths


{'2020-11-26': {'Quotes': [{'QuoteId': 1,
    'MinPrice': 50,
    'Direct': False,
    'OutboundLeg': {'CarrierIds': [1090],
     'OriginId': 43313,
     'DestinationId': 42414,
     'DepartureDate': '2020-11-26T00:00:00'},
    'QuoteDateTime': '2020-11-21T09:15:00'},
   {'QuoteId': 2,
    'MinPrice': 129,
    'Direct': True,
    'OutboundLeg': {'CarrierIds': [1218],
     'OriginId': 73097,
     'DestinationId': 42414,
     'DepartureDate': '2020-11-26T00:00:00'},
    'QuoteDateTime': '2020-11-24T17:16:00'}],
  'Carriers': [{'CarrierId': 1090, 'Name': 'Ryanair'},
   {'CarrierId': 1218, 'Name': 'Iberia'}],
  'Places': [{'Name': 'Barcelona',
    'Type': 'Station',
    'PlaceId': 42414,
    'IataCode': 'BCN',
    'SkyscannerCode': 'BCN',
    'CityName': 'Barcelona',
    'CityId': 'BARC',
    'CountryName': 'Spain'},
   {'Name': 'Paris Beauvais',
    'Type': 'Station',
    'PlaceId': 43313,
    'IataCode': 'BVA',
    'SkyscannerCode': 'BVA',
    'CityName': 'Paris',
    'CityId': 'PARI',
 

In [46]:
flights_dates = flight_days('Barcelona','London','2020-11-26','2020-12-10')

2020-11-26
Gathering fligths
2020-11-27
Gathering fligths
2020-11-28
Gathering fligths
2020-11-29
Gathering fligths
2020-11-30
Gathering fligths
2020-12-01
Gathering fligths
2020-12-02
Gathering fligths
2020-12-03
Gathering fligths
2020-12-04
Gathering fligths
2020-12-05
Gathering fligths
2020-12-06
Gathering fligths
2020-12-07
Gathering fligths
2020-12-08
Gathering fligths
2020-12-09
Gathering fligths
2020-12-10
Gathering fligths


In [47]:
flights_dates

{'2020-11-26': {'Quotes': [{'QuoteId': 1,
    'MinPrice': 107,
    'Direct': False,
    'OutboundLeg': {'CarrierIds': [1218],
     'OriginId': 42414,
     'DestinationId': 65698,
     'DepartureDate': '2020-11-26T00:00:00'},
    'QuoteDateTime': '2020-11-24T09:50:00'}],
  'Carriers': [{'CarrierId': 1218, 'Name': 'Iberia'}],
  'Places': [{'Name': 'Barcelona',
    'Type': 'Station',
    'PlaceId': 42414,
    'IataCode': 'BCN',
    'SkyscannerCode': 'BCN',
    'CityName': 'Barcelona',
    'CityId': 'BARC',
    'CountryName': 'Spain'},
   {'Name': 'London Heathrow',
    'Type': 'Station',
    'PlaceId': 65698,
    'IataCode': 'LHR',
    'SkyscannerCode': 'LHR',
    'CityName': 'London',
    'CityId': 'LOND',
    'CountryName': 'United Kingdom'}],
  'Currencies': [{'Code': 'USD',
    'Symbol': '$',
    'ThousandsSeparator': ',',
    'DecimalSeparator': '.',
    'SymbolOnLeft': True,
    'SpaceBetweenAmountAndSymbol': False,
    'RoundingCoefficient': 0,
    'DecimalDigits': 2}]},
 '2020-11-

Now let's create another function that returns the date of the cheapest flight.

In [48]:
#def lowest_price(origin, destination, start, end):
#    flights = flight_days(origin, destination, start, end)
#    prices = {date:min([quote["MinPrice"] for quote in flights[date]["Quotes"]]) for date in flights for quote in flights[date]["Quotes"]}
#    return min(prices, key = prices.get)

#flights = flight_prices('Barcelona','London','2020-11-26','2020-12-10')

#my_dict = {date1: [], date2: [],....}

my_dict = {}

for date in flights_dates: # We iterate over the dates
    print("The flights for this date are: ")
    print(flights_dates[date]['Quotes'])
    my_dict[date] = [] #We create an empty list of prices for the given date.
    for elem in flights_dates[date]['Quotes']: # We iterate over "Quotes" of each day.
        my_dict[date].append(elem['MinPrice']) # We fill the date with the prices.
    print("The prices for this date are: ",my_dict[date])
    print("The min price for this date is: ",min(my_dict[date])) # We determine the minimum price.

The flights for this date are: 
[{'QuoteId': 1, 'MinPrice': 107, 'Direct': False, 'OutboundLeg': {'CarrierIds': [1218], 'OriginId': 42414, 'DestinationId': 65698, 'DepartureDate': '2020-11-26T00:00:00'}, 'QuoteDateTime': '2020-11-24T09:50:00'}]
The prices for this date are:  [107]
The min price for this date is:  107
The flights for this date are: 
[{'QuoteId': 1, 'MinPrice': 97, 'Direct': False, 'OutboundLeg': {'CarrierIds': [1218], 'OriginId': 42414, 'DestinationId': 65698, 'DepartureDate': '2020-11-27T00:00:00'}, 'QuoteDateTime': '2020-11-24T17:16:00'}, {'QuoteId': 2, 'MinPrice': 172, 'Direct': True, 'OutboundLeg': {'CarrierIds': [1865], 'OriginId': 42414, 'DestinationId': 65655, 'DepartureDate': '2020-11-27T00:00:00'}, 'QuoteDateTime': '2020-11-24T09:50:00'}]
The prices for this date are:  [97, 172]
The min price for this date is:  97
The flights for this date are: 
[{'QuoteId': 1, 'MinPrice': 102, 'Direct': False, 'OutboundLeg': {'CarrierIds': [1218], 'OriginId': 42414, 'Destinati

The problem with this code is that it just gets the first date as the only one even if there are other days with the same price. Try to fix that in the lab.

In [166]:
min(my_dict.values())

[46, 85]

# Lab | Working with APIs

#### Instructions 

Following the class example, create a function that returns the price, names of origin and arrival airports and the name of the company. Do it for all the flights between two dates that cost the same.

In [137]:
cheap_dates={}
cheap_dates[date]=[]
for i in my_dict:
    if min(my_dict[i])==min(my_dict.values())[0]:
        cheap_dates[date].append(my_dict[i])

In [164]:
cheapest={}
cheapest[date]=[]
for i in cheap_dates:
    cheapest[date].append(min(cheap_dates[i][0]))

In [165]:
cheapest

{'2020-12-10': [46]}